source/components/hardware/hwregs.c

```diff
@@ -164,31 +164,56 @@ AcpiHwGetAccessBitWidth (
     ACPI_GENERIC_ADDRESS    *Reg,
     UINT8                   MaxBitWidth)
 {
+    UINT8                   AccessBitWidth;
 
-    if (!Reg->AccessWidth)
-    {
-        if (Reg->SpaceId == ACPI_ADR_SPACE_SYSTEM_IO)
-        {
-            MaxBitWidth = 32;
-        }
 
-        /*
-         * Detect old register descriptors where only the BitWidth field
-         * makes senses.
-         */
-        if (Reg->BitWidth < MaxBitWidth &&
-            !Reg->BitOffset && Reg->BitWidth &&
-            ACPI_IS_POWER_OF_TWO (Reg->BitWidth) &&
-            ACPI_IS_ALIGNED (Reg->BitWidth, 8))
-        {
-            return (Reg->BitWidth);
-        }
-        return (MaxBitWidth);
+    /*
+     * GAS format "register", used by FADT:
+     *  1. Detected if BitOffset is 0 and BitWidth is 8/16/32/64;
+     *  2. AccessSize field is ignored and BitWidth field is used for
+     *     determining the boundary of the IO accesses.
+     * GAS format "region", used by APEI registers:
+     *  1. Detected if BitOffset is not 0 or BitWidth is not 8/16/32/64;
+     *  2. AccessSize field is used for determining the boundary of the
+     *     IO accesses;
+     *  3. BitOffset/BitWidth fields are used to describe the "region".
+     *
+     * Note: This algorithm assumes that the "Address" fields should always
+     *       contain aligned values.
+     */
+    if (!Reg->BitOffset && Reg->BitWidth &&
+        ACPI_IS_POWER_OF_TWO (Reg->BitWidth) &&
+        ACPI_IS_ALIGNED (Reg->BitWidth, 8))
+    {
+        AccessBitWidth = Reg->BitWidth;
+    }
+    else if (Reg->AccessWidth)
+    {
+        AccessBitWidth = (1 << (Reg->AccessWidth + 2));
     }
     else
     {
-        return (1 << (Reg->AccessWidth + 2));
+        AccessBitWidth = ACPI_ROUND_UP_POWER_OF_TWO_8 (
+            Reg->BitOffset + Reg->BitWidth);
+    }
+
```

Comment:
IMO, the previous algorithm should be correct.
As it always treat AccessSize as good configuration.
Why should the BIOS developers provide us with a wrong AccessSize field?

On the other hand, we may not need to add ROUND_UP(BitOffset + BitWidth) line.
Because there may not be such a case.
As a "region" GAS, AccessSize should always be non-zero.
And BitOffset + BitWidth should be less than AccessSize.
Otherwise how can this GAS be a region descriptor?
So BIOS developer should have the duty to provide qualified "region" GAS to us.

That's why the previous commit.
However this bug forces me to change the code to this approach:
https://bugzilla.kernel.org/show_bug.cgi?id=151501

I'm still doubt if this approach is correct as I cannot reproduce the issue.
So we'll escalate this to ASWG to see if this is needed.

Before clarification, do not merge.